test/framework/src/main/java/org/elasticsearch/cluster/ESAllocationTestCase.java

```diff
@@ -268,6 +271,9 @@ public void applyFailedShards(RoutingAllocation allocation, List<FailedShard> fa
 
         @Override
         public void allocateUnassigned(RoutingAllocation allocation) {
+            if (this.skipAllocation) {
+                return;
```

Comment:
I don't understand this addition. As far as I can tell this makes this mock allocator behave quite differently from the real `GatewayAllocator` doesn't it? Your test only fails because of this difference in behaviour: if I comment this line out then your test passes without any changes to the production code. Can you provide a test case using the real allocator? I suggest adding to `DelayedAllocationIT` rather than here to ensure the test matches the production code more closely.

Reply:
I just want to simulate the case that if unassigned shard cannot be allocated in `GatewayAllocator`, then it also needs to be delayed in `ShardsAllocator`. Suppose we skip `GatewayAllocator` allocation, without this PR, if I remove the node that has replica shard, unassigned shard will be allocated immediately to the other node, with this PR, it will still be delayed until `delayed_timeout`. 

Considering the sequence based allocation decision could select correct node for allocation, so this PR will have side effects on phase 2 allocation that will delay shard allocation until _delayed_timeout_. For more infomation please reference https://github.com/elastic/elasticsearch/pull/46520#issuecomment-533178253.